bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/handler/LeapBridgeHandler.java

```diff
@@ -0,0 +1,763 @@
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Keystore/keystore password not configured");
+            return;
+        } else {
+            try (FileInputStream keystoreInputStream = new FileInputStream(config.keystore)) {
+                logger.trace("Initializing keystore");
+                KeyStore keystore = KeyStore.getInstance(KeyStore.getDefaultType());
+
+                keystore.load(keystoreInputStream, keystorePassword.toCharArray());
+
+                logger.trace("Initializing SSL Context");
+                KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+                kmf.init(keystore, keystorePassword.toCharArray());
+
+                TrustManager[] trustManagers;
+                if (config.certValidate) {
+                    // Use default trust manager which will attempt to validate server certificate from hub
+                    TrustManagerFactory tmf = TrustManagerFactory
+                            .getInstance(TrustManagerFactory.getDefaultAlgorithm());
+                    tmf.init(keystore);
+                    trustManagers = tmf.getTrustManagers();
+                } else {
+                    // Use no-op trust manager which will not verify certificates
+                    trustManagers = defineNoOpTrustManager();
+                }
+
+                sslContext = SSLContext.getInstance("TLS");
+                sslContext.init(kmf.getKeyManagers(), trustManagers, null);
+
+                sslsocketfactory = sslContext.getSocketFactory();
+            } catch (FileNotFoundException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Keystore file not found");
+                return;
+            } catch (CertificateException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Certificate exception");
+                return;
+            } catch (UnrecoverableKeyException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Key unrecoverable with supplied password");
+                return;
+            } catch (KeyManagementException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Key management exception");
+                logger.debug("Key management exception", e);
+                return;
+            } catch (KeyStoreException | NoSuchAlgorithmException | IOException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Error initializing keystore");
+                logger.debug("Error initializing keystore", e);
+                return;
+            }
+        }
+
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Connecting");
+        scheduler.submit(this::connect); // start the async connect task
```

Comment:
Since this is asynchronous, if this handler is disposed immediately after getting initialized there is a chance it could get disposed before the connection finishes which would result in an illegal state.

The only slow operation you are performing as part of `connect` is the creation of the SSLSocket. Perhaps you could focus on just making that asynchronous instead.
Another option would be to cancel the future handle interrupt within your `connect` method.

Reply:
>Since this is asynchronous, if this handler is disposed immediately after getting initialized there is a chance it could get disposed before the connection finishes which would result in an illegal state.

That’s a good point. I think most bindings that open some sort of network connection use pretty much the same pattern, though. If there is a real possibility of dispose() being called before the async part of initialization is done, maybe the framework should be modified to keep that from happening, or some recommendation should be made in the docs about how best to deal with it?

>The only slow operation you are performing as part of connect is the creation of the SSLSocket. Perhaps you could focus on just making that asynchronous instead.

Well, connect() doesn’t do much before creating the the SSL socket and starting the handshake, so I don’t think that would help too much. Even after opening the connection, it just starts a couple of threads and queues a few commands for sending.

>Another option would be to cancel the future handle interrupt within your connect method.

I’m not sure what you mean by this. I could save the Future returned by submit and attempt to call cancel on it in dispose(). But are createSocket() and startHandshake() even interruptible? And I assume dispose() would need to wait for connect() to exit before continuing to clean up, but I’m not sure what the best way to do that is.

Reply:
>  But are createSocket() and startHandshake() even interruptible? And I assume dispose() would need to wait for connect() to exit before continuing to clean up, but I’m not sure what the best way to do that is.

You don't necessary have to interrupt them so much as make sure that any results are silently discarded. You might have to do some synchronous blocks on your asynchronous stuff to make sure that a dispose() followed by an initialize() wouldn't cause some sort of collision on resource access (if it is even possible for such a collision).

Reply:
I just pushed a commit that should take care of this.